### PR TITLE
[IE CLDNN] Improve Gather performance and add fusing support

### PR DESCRIPTION
**Pull request by sshlyapn**
Added support for Fused Ops mechanism to Gather primitive and changed indices calculation which improves performance up to 9 times

JIRA: 32376

### CODE REVIEW

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp [hunk 1]
```diff
@@ -46,48 +47,86 @@ ParamsKey GatherKernelRef::GetSupportedKey() const {
...
     k.EnableBatching();
     k.EnableDifferentTypes();
     return k;
 }
 
-static size_t getPartSize(const gather_params& params, int32_t axis) {
-    size_t partSize = 1;
-    for (size_t i = params.inputs[0].Dimentions() - axis; i > 0; --i) partSize *= params.inputs[0].GetDims()[i - 1].v;
-    return partSize;
+static size_t GetNonEmpyDimsNumber(const DataTensor& data_tensor) {
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp:61]: Maybe, GetNonEmptyDimsNumber(...) ?

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/gather_ref.cl [hunk 2]
```diff
@@ -15,19 +15,38 @@
 
 #include "include/include_all.cl"
 
-KERNEL(gather_ref)(const __global INPUT0_TYPE* dictionary, const __global INPUT1_TYPE* indices, __global OUTPUT_TYPE* output)
+#define INPUT_AXIS_INDEX (uint)indices[indeces_idx]
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/gather_ref.cl:18]: indices[indices_idx]

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp (hunk 1 above)
Comment (sshlyapn) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/gather/gather_kernel_ref.cpp]: Right, thank you

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/gather_ref.cl (hunk 2 above)
Comment (sshlyapn) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/gather_ref.cl]: Fixed

**Pull request merged**